File: solucoes/python/atividade_central_impressao.py
"""Solucao de referencia da atividade da Aula 2."""


class Pilha:
    def __init__(self, capacidade=20):
        self._dados = []
        self._capacidade = capacidade

    def esta_vazia(self):
        return not self._dados

    def empilhar(self, valor):
        if len(self._dados) == self._capacidade:
            raise OverflowError("pilha cheia")
        self._dados.append(valor)

    def desempilhar(self):
        if self.esta_vazia():
            raise IndexError("pilha vazia")
        return self._dados.pop()

    def consultar_topo(self):
        if self.esta_vazia():
            raise IndexError("pilha vazia")
        return self._dados[-1]

    def valores_do_topo_para_base(self):
        return list(reversed(self._dados))


class NoFila:
    def __init__(self, valor):
        self.valor = valor
        self.proximo = None


class Fila:
    def __init__(self):
        self.inicio = None
        self.fim = None

    def esta_vazia(self):
        return self.inicio is None

    def enfileirar(self, valor):
        novo = NoFila(valor)
        if self.fim is None:
            self.inicio = novo
        else:
            self.fim.proximo = novo
        self.fim = novo

    def desenfileirar(self):
        if self.esta_vazia():
            raise IndexError("fila vazia")
        removido = self.inicio
        self.inicio = removido.proximo
        if self.inicio is None:
            self.fim = None
        return removido.valor

    def consultar_frente(self):
        if self.esta_vazia():
            raise IndexError("fila vazia")
        return self.inicio.valor

    def valores(self):
        resultado = []
        atual = self.inicio
        while atual is not None:
            resultado.append(atual.valor)
            atual = atual.proximo
        return resultado


def processar_proximo(espera, historico):
    if espera.esta_vazia():
        return False
    trabalho = espera.desenfileirar()
    historico.empilhar(trabalho)
    print(f"Trabalho processado: {trabalho}")
    return True


def main():
    espera = Fila()
    historico = Pilha()
    for trabalho in (101, 102, 103):
        espera.enfileirar(trabalho)
    processar_proximo(espera, historico)
    processar_proximo(espera, historico)
    espera.enfileirar(104)
    print("Fila de espera:", *espera.valores())
    print("Historico - topo para base:", *historico.valores_do_topo_para_base())
    print("Proximo trabalho:", espera.consultar_frente())
    print("Ultimo trabalho processado:", historico.consultar_topo())


if __name__ == "__main__":
    main()

File: aula02-pilhas-filas/solucoes/c/atividade_central_impressao.c
#include <stdio.h>
#include <stdlib.h>

#define CAPACIDADE 20

typedef struct {
    int dados[CAPACIDADE];
    int topo;
} Pilha;

typedef struct NoFila {
    int valor;
    struct NoFila *proximo;
} NoFila;

typedef struct {
    NoFila *inicio;
    NoFila *fim;
} Fila;

void pilha_criar(Pilha *p) { p->topo = -1; }
int pilha_esta_vazia(const Pilha *p) { return p->topo == -1; }

int pilha_empilhar(Pilha *p, int valor) {
    if (p->topo == CAPACIDADE - 1) return 0;
    p->dados[++p->topo] = valor;
    return 1;
}

int pilha_desempilhar(Pilha *p, int *valor) {
    if (pilha_esta_vazia(p)) return 0;
    if (valor != NULL) *valor = p->dados[p->topo];
    p->topo--;
    return 1;
}

int pilha_consultar_topo(const Pilha *p, int *valor) {
    if (pilha_esta_vazia(p)) return 0;
    if (valor != NULL) *valor = p->dados[p->topo];
    return 1;
}

void pilha_exibir(const Pilha *p) {
    for (int i = p->topo; i >= 0; i--) {
        printf("%d%s", p->dados[i], i > 0 ? " " : "");
    }
    printf("\n");
}

void fila_criar(Fila *f) { f->inicio = NULL; f->fim = NULL; }
int fila_esta_vazia(const Fila *f) { return f->inicio == NULL; }

int fila_enfileirar(Fila *f, int valor) {
    NoFila *novo = malloc(sizeof(NoFila));
    if (novo == NULL) return 0;
    novo->valor = valor;
    novo->proximo = NULL;
    if (f->fim == NULL) f->inicio = novo;
    else f->fim->proximo = novo;
    f->fim = novo;
    return 1;
}

int fila_desenfileirar(Fila *f, int *valor) {
    if (fila_esta_vazia(f)) return 0;
    NoFila *removido = f->inicio;
    if (valor != NULL) *valor = removido->valor;
    f->inicio = removido->proximo;
    if (f->inicio == NULL) f->fim = NULL;
    free(removido);
    return 1;
}

int fila_consultar_frente(const Fila *f, int *valor) {
    if (fila_esta_vazia(f)) return 0;
    if (valor != NULL) *valor = f->inicio->valor;
    return 1;
}

void fila_exibir(const Fila *f) {
    for (NoFila *atual = f->inicio; atual != NULL; atual = atual->proximo) {
        printf("%d%s", atual->valor, atual->proximo != NULL ? " " : "");
    }
    printf("\n");
}

void fila_destruir(Fila *f) {
    int ignorado;
    while (fila_desenfileirar(f, &ignorado)) { }
}

int processar_proximo(Fila *espera, Pilha *historico) {
    int trabalho;
    if (!fila_desenfileirar(espera, &trabalho)) return 0;
    if (!pilha_empilhar(historico, trabalho)) return 0;
    printf("Trabalho processado: %d\n", trabalho);
    return 1;
}

int main(void) {
    Fila espera;
    Pilha historico;
    int valor;
    fila_criar(&espera);
    pilha_criar(&historico);
    fila_enfileirar(&espera, 101);
    fila_enfileirar(&espera, 102);
    fila_enfileirar(&espera, 103);
    processar_proximo(&espera, &historico);
    processar_proximo(&espera, &historico);
    fila_enfileirar(&espera, 104);
    printf("Fila de espera: "); fila_exibir(&espera);
    printf("Historico - topo para base: "); pilha_exibir(&historico);
    fila_consultar_frente(&espera, &valor);
    printf("Proximo trabalho: %d\n", valor);
    pilha_consultar_topo(&historico, &valor);
    printf("Ultimo trabalho processado: %d\n", valor);
    fila_destruir(&espera);
    return 0;
}
